cg/smart_contracts/ClearGDPR.sol

```diff
@@ -1,153 +1,163 @@
+        processors = new address[](_newProcessors.length + 1);
+        processors[0] = controller;
+        for(uint256 i = 0; i < _newProcessors.length; i++){
+            //requires that none of the processors is equal to the controller
+            //Processors should all be different
+            processors[i + 1] = _newProcessors[i];
+        }
         return true;
-      }
     }
-    return false;
-  }
   
-  function areAllValidProcessors(address[] _processors) public view returns(bool){
-    for(uint256 i = 0; i < _processors.length; i++){
-      require(isProcessor(_processors[i]));
+    function isProcessor(address _processor) public view returns(bool){
+        for(uint256 i = 0; i < processors.length; i++){
+            if(processors[i] == _processor){
+                return true;
+            }
+        }
+        return false;
     }
-    return true;
-  }
   
-  function recordProcessorsUpdate(address[] _newProcessors) public onlyController returns(bool){
-    require(setProcessors(_newProcessors));
-    emit Controller_ProcessorsUpdated(_newProcessors);  
-    return true;
-  }
+    function areAllValidProcessors(address[] _processors) public view returns(bool){
+        for(uint256 i = 0; i < _processors.length; i++){
+            require(isProcessor(_processors[i]));
+        }
+        return true;
+    }
+  
+    function recordProcessorsUpdate(address[] _newProcessors) public onlyController returns(bool){
+        require(setProcessors(_newProcessors));
+        emit Controller_ProcessorsUpdated(_newProcessors);  
+        return true;
+    }
 
-  function recordConsentGivenTo(bytes32 _subjectIdHash, address[] _processorsWhiteListedBySubject) public onlyController notErased(_subjectIdHash)  returns (bool){
-    require(areAllValidProcessors(_processorsWhiteListedBySubject));
-    for(uint256 i = 0; i < processors.length; i++){
-      subjects[_subjectIdHash].statePerProcessor[processors[i]] = State.notShareable;  
-    }
-    for(i = 0; i < _processorsWhiteListedBySubject.length; i++){
-      subjects[_subjectIdHash].statePerProcessor[_processorsWhiteListedBySubject[i]] = State.shareable;  
-    }
-    subjects[_subjectIdHash].isErased = false;
-    emit Controller_ConsentGivenTo(_subjectIdHash, _processorsWhiteListedBySubject);
-    return true;
-  }
+    function recordConsentGivenTo(bytes32 _subjectId, address[] _processorsConsented) public onlyController notErased(_subjectId) returns(bool){
+        require(areAllValidProcessors(_processorsConsented));
+        for(uint256 i = 0; i < processors.length; i++){
+            subjects[_subjectId].statePerProcessor[processors[i]] = State.notShareable;  
+        }
+        for(i = 0; i < _processorsConsented.length; i++){
+            subjects[_subjectId].statePerProcessor[_processorsConsented[i]] = State.shareable;  
+        }
+        subjects[_subjectId].statePerProcessor[controller] = State.shareable;  
```

Comment:
the controller always has consent. I think the state shareable is not the best name for when a Quorum node has consent. We should probably rename it, because the controller can't actually share the subject's data without the subject's consent for a specific processor. Suggestion: unconsented, consented, erased, objected

Reply:
I agree. Will require refactoring of consts in JS codebase.